### ISSUE TITLE
master_june24: wrong/bugged and unnecessary SIMD handling of numerators and denominators 

### ISSUE THREAD

**Issue by valassi** [madgraph5/madgraph4gpu#894]
Another issue introduced in #830 and being reviewed in #882.

The SIMD handling of numersators and denominators is conceptually wrong, and a potetial source of bugs.

https://github.com/valassi/madgraph4gpu/blob/8e312bc02d9d072615fcb1052b5db54754498517/epochX/cudacpp/gg_tt.mad/SubProcesses/P1_gg_ttx/CPPProcess.cc#L349
```
        for( int i = 0; i < neppV; ++i )
        {
          if( channelids_sv[i] == 1 ) numerators_sv += cxabs2( amp_sv[0] );
          denominators_sv += cxabs2( amp_sv[0] );
        }
```

The problem here is that *inside the loop over SIMD elements, the whole SIMD vector is used*. This is conceptually completely wrong. There would be two correct ways of doing this
- Either with an explicit loop, but then only `numerators_sv[i]` and `denominators_sv[i]` would need to be manipulated
- Or, using SIMD notation, one can avoid the explicit loop and manipulate only SIMD vectors (much faster, if you do not want to lose SIMD performance)

The second approach is straightforward for denominators. There is no need for an explicit loop.

The second approach is also better for numerators, If one assumes that all channels in the SIMD vector are equal (I am quite sure we had discussed this, I will dig it up... I will read this for instance https://github.com/madgraph5/madgraph4gpu/issues/765#issuecomment-1853672838), then a single channelId is enough for all elements in the vector. In the case (which I think we avoid, wth the warp mechanism) that different SIMD elements refer to different channels, a mask can be used to preserve SIMD performance. This was discussed at length (and I thought we had agreed....) in https://github.com/madgraph5/madgraph4gpu/issues/765#issuecomment-1931499184

The current implementation is also a potential source of bugs. Specifically, if there were different channelids in the same SIMD vector, then I suspect the code would give wrong results (maybe I am wrong). My impression is that currently the code 'magically' or 'by mistake' gives the right result, because all channels in the same SIMD vector are equal. For a SIMD vector of 4 elements, the code above effectively sums the correct quantities to numerators and denominators 4 times, instead of only once? Since what we care about is only numerator/denominator, then `num/den=(4*num)/(4*den)` and this works. But this is a lucky coincidence. The code must be fixed.

**Comment by roiser** [comment]
Please note that the channelIds as provided by Fortran will be the same for a warp (or SIMD array), this was discussed and agreed when we prepared for this work.

My recollection of the discussion on masking was that we could not use it because we handle two different fundamental types (int and uint) and there was no SIMD operation which could handle this.

**Comment by oliviermattelaer** [comment]
```
If the assumption 'all channelids are the same' holds
```

Yes, the assumption SHOULD/MUST hold (and checking it with #898 is the way to go).
 

Concerning
```
        for( int i = 0; i < neppV; ++i )
        {
          if( channelids_sv[i] == 1 ) numerators_sv[i] += cxabs2( amp_sv[0] );
          denominators_sv += cxabs2( amp_sv[0] );
        }
```
I have discussed with @roiser and indeed we are repeating the same computation and we could simplify the work into the following lines:
```
          if( channelids_sv[0] == 1 ) numerators_sv += cxabs2( amp_sv[0] );
          denominators_sv += cxabs2( amp_sv[0] );
```

Now one issue which is not clear to me here is how those lines handle the case where `nb_warp=2`?
@valassi does the code goes twice into those lines (with different channelids_sv) and then the above is ok?

Thanks @valassi for spotting this, this could indeed be problematic
(and thanks @roiser for testing so much combination here)

**Comment by valassi** [comment]
> ```
> If the assumption 'all channelids are the same' holds
> ```
> 
> Yes, the assumption SHOULD/MUST hold (and checking it with #898 is the way to go).

Thanks Olivier :-)

Yes this is now checked in 898 for the first SIMD vector. Snd for mixed case there is an additional check needed in #924, also completed.

> Concerning
> 
> ```
>         for( int i = 0; i < neppV; ++i )
>         {
>           if( channelids_sv[i] == 1 ) numerators_sv[i] += cxabs2( amp_sv[0] );
>           denominators_sv += cxabs2( amp_sv[0] );
>         }
> ```
> 
> I have discussed with @roiser and indeed we are repeating the same computation and we could simplify the work into the following lines:
> 
> ```
>           if( channelids_sv[0] == 1 ) numerators_sv += cxabs2( amp_sv[0] );
>           denominators_sv += cxabs2( amp_sv[0] );
> ```

Well. This is what the upstream/master code was *already* doing. (With the additional advantage that the same code works for SIMD and no SIMD or GPU, because a scalar channelId is enough - and I can use a scalar channelId *precisely* because I have now implemented the sanity check 898). (And with the additional handling of the channelId=0 - not nullptr - for helicity filtering).
https://github.com/madgraph5/madgraph4gpu/blob/606ee3bd25a9dd9f452094265116d5dcd20304ac/epochX/cudacpp/gg_tt.mad/SubProcesses/P1_gg_ttx/CPPProcess.cc#L335
```
#ifdef MGONGPU_SUPPORTS_MULTICHANNEL
      if( channelId == 1 ) numerators_sv += cxabs2( amp_sv[0] );
      if( channelId != 0 ) denominators_sv += cxabs2( amp_sv[0] );
#endif
```

And what the code is doing again now
https://github.com/valassi/madgraph4gpu/blob/4bf4e1f358557f3d7f80edd3097a52191e819a3f/epochX/cudacpp/gg_tt.mad/SubProcesses/P1_gg_ttx/CPPProcess.cc#L337
```
#ifdef MGONGPU_SUPPORTS_MULTICHANNEL
      if( channelId == 1 ) numerators_sv += cxabs2( amp_sv[0] );
      if( channelId != 0 ) denominators_sv += cxabs2( amp_sv[0] );
#endif
```

That is to say, in practice I have *undone* practically all changes in calculate_wavefunctions, as those changes were wrong/buggy and unnecessary/useless.

> Now one issue which is not clear to me here is how those lines handle the case where `nb_warp=2`? @valassi does the code goes twice into those lines (with different channelids_sv) and then the above is ok?

Good question, not so much about nb_warp=2 but about mixed type (for any even number of warps). For mixed type, the design was in upstream/master (and I kept exactly the same after my undoing 830) that calculate_wavefunction internall loops over iParity=0,1 for mixed mode. 

Then the question is what we agreed to move in #926. You have two options
- You check elsewhere that two consecutive vectors have the samne channelId, and you use a single channelId. This is what I am doing now and what will be in 882, as we agreed to defer any changes to later.
- OR, you pass two separate channelIds in mixed mode, and you use the first for iParity=0 and the second for iParity=1. This is not rocker science complex, but is some work, an din my opinion not necessary as a use case. But lets defer  the discussion to 926. Here I just explained the technical mechanisms.

> Thanks @valassi for spotting this, this could indeed be problematic 

Thanks to you Olivier. This took me quite some additional testing with respect to what was in 830, but I think it's more robust now.

As discussed at length above, this is fixed in #882. Linking it there and closing this issue.

**Comment by valassi** [comment]
and closing after linking
